containers-toolkit/Private/CommonToolUtilities.psm1

```diff
@@ -29,15 +29,48 @@ class FileDigest {
     }
 }
 
+class FileDownloadParameters {
+    [ValidateSet("Containerd", "Buildkit", "nerdctl", "WinCNIPlugin")]
+    [string]$Feature
+    [string]$Repo
+    [string]$Version = "latest"
+    [ValidateSet("386", "amd64", "arm64", "arm")]
+    [string]$OSArchitecture = "$env:PROCESSOR_ARCHITECTURE"
+    [string]$DownloadPath = "$HOME\Downloads"
+    [string]$ChecksumSchemaFile
+    [string]$FileFilterRegEx
+
+    FileDownloadParameters(
+        [string]$feature,
+        [string]$repo,
+        [string]$version = "latest",
+        [string]$arch = "$env:PROCESSOR_ARCHITECTURE",
+        [string]$downloadPath = "$HOME\Downloads",
+        [string]$checksumSchemaFile = $null,
+        [string]$fileFilterRegEx = $null
+    ) {
+        $this.Feature = $feature
+        $this.Repo = $repo
+        $this.Version = $version
+        $this.OSArchitecture = $arch
+        $this.DownloadPath = $downloadPath
+        $this.ChecksumSchemaFile = $checksumSchemaFile
+        $this.FileFilterRegEx = $fileFilterRegEx
+    }
+}
+
+
 Add-Type @'
 public enum ActionConsent {
     Yes = 0,
     No = 1
 }
 '@
 
-$VALID_HASH_FUNCTIONS = @("SHA1", "SHA256", "SHA384", "SHA512", "MD5")
-
+$HASH_FUNCTIONS = @("SHA1", "SHA256", "SHA384", "SHA512", "MD5")
+$HASH_FUNCTIONS_STR = $HASH_FUNCTIONS -join '|' # SHA1|SHA256|SHA384|SHA512|MD5
+$NERDCTL_CHECKSUM_FILE_PATTERN = "(?<hashfunction>(?:^({0})))" -f ($HASH_FUNCTIONS -join '|')
```

Comment:
can this be generalized in any way as opposed to being `nerdctl` specific? How many checksum patterns are there between the current 3 tools?

Reply:
Checksum naming formats:

- Containerd: <release-file-name>.tar.gz.sha256sum (e.g., containerd-1.7.22-windows-amd64.tar.gz.sha256sum)
- Buildkit: <release-file-name>.sbom.json (e.g., buildkit-v0.16.0.windows-amd64.sbom.json)
- nerdctl: SHA256SUMS (the checksum file is simply named SHA256SUMS)

We filter based on (1) the operating system (e.g., Windows) and (2) architecture (e.g., amd64, arm64). After filtering, we match the binary archive file (e.g., containerd-1.7.22-windows-amd64.tar.gz) with its corresponding checksum file (e.g., containerd-1.7.22-windows-amd64.tar.gz.sha256sum).

However, for nerdctl, this process isn't feasible due to the generic naming of the checksum file. Therefore, we must handle nerdctl with a specific pattern. Although we could pass this as an additional filter from the Install-Nerdctl function to this function, we prefer to lock it down, ensuring that only specific files can be downloaded with this package.